d2/src/main/java/com/linkedin/d2/xds/XdsToD2PropertiesAdaptor.java

```diff
@@ -154,6 +187,25 @@ public void listenToService(String serviceName)
     });
   }
 
+  private static boolean isSymlinkNode(String nodeNameOrPath)
+  {
+    return nodeNameOrPath != null && nodeNameOrPath.indexOf(SYMLINK_NODE_IDENTIFIER) >= 0;
+  }
+
+  private void listenToSymlink(String name, String fullResourceName)
+  {
+    // use full resource name ("/d2/clusters/$FooClusterMater", "/d2/uris/$FooClusterMaster") as the key
+    // instead of just the symlink name ("$FooClusterMaster") to differentiate clusters and uris symlink resources.
+    _watchedSymlinkResources.computeIfAbsent(fullResourceName, k ->
```

Comment:
The results are stored without the cluster and uri prefixes in the bi-map later:
```

  // Mapping between a symlink name, like "$FooClusterMaster" and the actual node name it's pointing to, like
  // "FooCluster-prod-ltx1".
  // (Note that this name does NOT include the full path so that it works for both cluster symlink
  // "/d2/clusters/$FooClusterMaster" and uri-parent symlink "/d2/uris/$FooClusterMaster").
  private final HashBiMap<String, String> _symlinkAndActualNode = HashBiMap.create();
```

Should the symlinks for uris and clusters happen to be different, they will overwrite each other and be quite error prone. For consistency, the symlink resource names should be kept the same here as well, or they should be kept separately in `_symlinkAndActualNode` too.

Reply:
symlinks for uris and clusters are always the same. The tooling always update the URI and cluster symlinks together.